crates/chain/src/keychain/txout_index.rs

```diff
@@ -11,6 +11,75 @@ use core::{fmt::Debug, ops::Deref};
 
 use crate::Append;
 
+/// Represents updates to the derivation index of a [`KeychainTxOutIndex`].
+/// It maps each keychain `K` to a descriptor and its last revealed index.
+///
+/// It can be applied to [`KeychainTxOutIndex`] with [`apply_changeset`]. [`ChangeSet] are
+/// monotone in that they will never decrease the revealed derivation index.
+///
+/// [`KeychainTxOutIndex`]: crate::keychain::KeychainTxOutIndex
+/// [`apply_changeset`]: crate::keychain::KeychainTxOutIndex::apply_changeset
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "K: Ord + serde::Deserialize<'de>",
+            serialize = "K: Ord + serde::Serialize"
+        )
+    )
+)]
+#[must_use]
+pub struct ChangeSet<K>(pub BTreeMap<K, (Descriptor<DescriptorPublicKey>, u32)>);
+
+impl<K> ChangeSet<K> {
+    /// Get the inner map of the keychain to its new derivation index.
+    pub fn as_inner(&self) -> &BTreeMap<K, (Descriptor<DescriptorPublicKey>, u32)> {
+        &self.0
+    }
+}
+
+impl<K: Ord> Append for ChangeSet<K> {
+    /// Append another [`ChangeSet`] into self.
+    ///
+    /// For each keychain in the given [`ChangeSet`]:
+    /// - If the keychain already exists in the old [`ChangeSet`], and the two descriptors are the
+    /// same, we increase the index when new index > old index.
+    /// - If the keychain already exists, but with a different descriptor, we overwrite the
+    /// original descriptor with the changeset one.
```

Comment:
I'm wondering whether it is a good idea to allow changing the descriptor for a keychain... It may result in some potentially confusing behavior? The other option is to ignore it, and add a `debug_assert!` statement when a changeset attempts to change the descriptor.

Reply:
Let's make this monotone. Just take the highest `(descriptor, index)` pair as a tiebreak (`Descriptor` implements `Ord`). Don't allow changing the descriptor for a given in the API so hopefully this can't happen in practice but we at least make the changesets monotone in case it does.